src/nvim/cursor_shape.c

```diff
@@ -36,11 +38,61 @@ static cursorentry_T shape_table[SHAPE_IDX_COUNT] =
   {0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR},
 };
 
-/*
- * Parse the 'guicursor' option ("what" is SHAPE_CURSOR) or 'mouseshape'
- * ("what" is SHAPE_MOUSE).
- * Returns error message for an illegal option, NULL otherwise.
- */
+
+Dictionary cursor_shape_dict(void)
+{
+  /* Array args = ARRAY_DICT_INIT; */
+  Dictionary all = ARRAY_DICT_INIT;
+  char buf[30] = "uninitialized";
+
+  /* shape_table */
+  for (int i = 0; i < SHAPE_IDX_COUNT; i++) {
+
+    ILOG("Adding entry");
+    Dictionary dic = ARRAY_DICT_INIT;
+    cursorentry_T *cur = &shape_table[i];
+    /* I noticed msgp*/
+    /* switch (cur->shape) { */
+    /*   case SHAPE_BLOCK: STRCPY(buf, "block"); */
+    /*   case SHAPE_VER: STRCPY(buf, "vertical"); */
+    /*   case SHAPE_HOR: STRCPY(buf, "horizontal"); */
+    /*   default:; */
+    /* }; */
+    /* PUT(dic, "shape", STRING_OBJ(buf)); */
+    PUT(dic, "shape", INTEGER_OBJ(cur->shape));
+    PUT(dic, "mshape", INTEGER_OBJ(cur->mshape));
+    PUT(dic, "percentage", INTEGER_OBJ(cur->percentage));
+    PUT(dic, "blinkwait", INTEGER_OBJ(cur->blinkwait));
+    PUT(dic, "blinkon", INTEGER_OBJ(cur->blinkon));
+    PUT(dic, "blinkoff", INTEGER_OBJ(cur->blinkoff));
+    PUT(dic, "id", INTEGER_OBJ(cur->id));
+    PUT(dic, "id_lm", INTEGER_OBJ(cur->id_lm));
+    PUT(dic, "name", STRING_OBJ(cur->name));
+
+    PUT(dic, "used_for", INTEGER_OBJ(cur->used_for));
+
+    /* if (attrs.special != -1) { */
+    /*   PUT(hl, "special", INTEGER_OBJ(attrs.special)); */
+    /* v */
+ 
+    /* PUT(dic, "name", STRING_OBJ(cur->name)); */
+    PUT(all, cur->name, DICTIONARY_OBJ(dic));
```

Comment:
Should we pass this over an array of "cursorentry" or a dict of cursorentry with the mode as the key ?